module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -458,45 +471,56 @@ public function getHolding($bibId, array $patron = null, array $options = [])
      */
     public function patronLogin($username, $password)
     {
-        // Get user id
-        $query = ['query' => 'username == ' . $username];
-        $response = $this->makeRequest('GET', '/users', $query);
+        if ($this->config['User']['okapi_login'] ?? false) {
+            $tenant = $this->config['API']['tenant'];
+            $credentials = compact('tenant', 'username', 'password');
+            // Get token
+            try {
+                $response = $this->makeRequest(
+                    'POST',
+                    '/authn/login',
+                    json_encode($credentials)
+                );
+                $debugMsg = 'User logged in. User: ' . $username . '.';
+                // We've authenticated the user with Okapi, but we only have their
+                // username; set up a query to retrieve full info below.
+                $query = 'username == ' . $username;
+                // Replace admin with user as tenant if configured to do so:
+                if ($this->config['User']['use_user_token'] ?? false) {
+                    $this->token = $response->getHeaders()->get('X-Okapi-Token')
+                        ->getFieldValue();
+                    $debugMsg .= ' Token: ' . substr($this->token, 0, 30) . '...';
+                }
+                $this->debug($debugMsg);
+            } catch (Exception $e) {
+                return null;
+            }
+        } else {
+            // Construct user query using barcode, username, etc.
+            $usernameField = $this->config['User']['username_field'] ?? 'username';
+            $passwordField = $this->config['User']['password_field'] ?? 'password';
+            $query = $usernameField . ' == ' . $this->escapeCql($username);
+            if (!empty($passwordField)) {
+                $query .= " and {$passwordField} == " . $this->escapeCql($password);
+            }
+        }
+
+        $response = $this->makeRequest('GET', '/users', compact('query'));
         $json = json_decode($response->getBody());
-        if (count($json->users) == 0) {
-            throw new ILSException("User not found");
+        if (count($json->users) !== 1) {
```

Comment:
I think it's a good idea to be more specific in the test here; hopefully this will work!